File: solutions/python/simple-linked-list/1/simple_linked_list.py
class Element:
    def __init__(self, datum, _next=None):
        self.datum = datum
        self._next = _next

    def next(self):
        return self._next

    def value(self):
        return self.datum


class LinkedList:
    def __init__(self):
        self._head = None
        self._size = 0

    def __len__(self):
        return self._size

    def head(self):
        if self._head is None:
            raise EmptyListException("List is empty")
        return self._head

    def push(self, datum):
        self._head = Element(datum, self._head)
        self._size += 1

    def pop(self):
        if self._head is None:
            raise EmptyListException("List is empty")
        datum = self._head.datum
        self._head = self._head._next
        self._size -= 1
        return datum

    def reverse(self):
        prev, curr = None, self._head
        while curr:
            nxt = curr._next
            curr._next = prev
            prev = curr
            curr = nxt
        self._head = prev


class EmptyListException(Exception):
    pass

File: solutions/ruby/simple-linked-list/1/simple_linked_list.rb
class Element
  attr_reader :datum
  attr_accessor :next

  def initialize(datum)
    @datum = datum
    @next = nil
  end
end

class SimpleLinkedList
  def initialize(array = nil)
    @head = nil
    if array
      array.to_a.each { |e| push(Element.new(e)) }
    end
  end

  def push(element)
    element.next = @head
    @head = element
    self
  end

  def pop
    return nil unless @head
    element = @head
    @head = @head.next
    element
  end

  def to_a
    result = []
    current = @head
    while current
      result << current.datum
      current = current.next
    end
    result
  end

  def reverse!
    prev = nil
    current = @head
    while current
      nxt = current.next
      current.next = prev
      prev = current
      current = nxt
    end
    @head = prev
    self
  end
end
